keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2862 @@
+Constraints](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#resource-constraints), which provides:
+
+- CEL estimated cost limits
+- CEL runtime cost limits
+- Go context cancelation as a way of halting CEL execution if the request
+  context is canceled for any reason.
+
+ Estimated cost is, unfortunately, not something we can offer for admission with
+ any kind of guarantees attached due to the already listed issues that have
+ prevented use from enforcing type safety. We could instead compute estimated
+ for the same cases where we provide informational type checking, in which case
+ we can report any cost limit violations in the same way we report type checking
+ violations. Note that for built-in types, where `max{Length,Items,Properties}`
+ value valiations are not available, estimated cost calculations will not be
+ nearly as helpful or actionable. I recommend we do not attempt any estimated
+ cost calculations on built-in types until the value validations are available.
+
+ Runtime cost limits can should be established and enforced. Exceeding the cost
+ limit will trigger the `FailurePolicy`, so this will need to be documented, but
+ unlike webhooks, runtime cost is deterministic (it is purely a function of the
+ input data and the CEL expression and is independent of underlying hardware or
+ system load), making it less of a concern for control plane availability than
+ webhook timeouts.
+
+ The request's Go context will be passed in to all CEL evaluations such that
+ cancelation halts CEL evaluation, if, for any reason, the context is canceled.
+
+#### Safety Features
+
+<<[UNRESOLVED jpbetz, ?? ]>>
+Decide exactly what we are willing to commit to here.
+<<[/UNRESOLVED]>>
+
+- Configurable admission blocking write requests made internally in
+  kube-apiserver during server startup (like RBAC default policy reconciliation)
+  making it impossible for a server to start up healthy. (This is not specific to CEL?)
+- Ability to skip specific resource types - Admission Controller Webhook
+  configuration rule cannot exclude specific resources:
+  https://github.com/kubernetes/kubernetes/issues/92157
+
+#### Aggregated API servers
+
+Main complications (provided by @liggitt):
+
+- The API server validating/persisting the ValidatingAdmissionPolicy
+  instances isn't the same one serving aggregated types, so wouldn't necessarily
+  have schema info to check type safety.
+
+- The aggregated API server is responsible for enforcing admission on its custom
+  types, so the implementation that reads ValidatingAdmissionPolicy
+  instances and enforces them would have to live in k8s.io/apiserver and be
+  active in aggregated API servers to enforce admission on aggregated types
+  effectively (same as admission webhooks today).
+
+Plan:
+
+- Do not offer type checking for aggregated types.
+- Support ValidatingAdmissionPolicy in aggregated API servers.
+
+
+#### Short Circuiting
```

Comment:
I think you should wait for a clear example use case before considering this.

Reply:
SGTM. I'll shelve this one.